### PR TITLE
AIP-5879 Admission Webhook 1.22 Compatibility

### PR DESCRIPTION
**Pull request by aaron-arellano**
Updates the  admission webhook to be compatible with kubernetes v1.22 per issue #6428.

### CODE REVIEW

File: components/admission-webhook/main.go [hunk 1]
```diff
@@ -484,7 +483,7 @@ func mutatePods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
 
 	crdclient := getCrdClient()
 	list := &settingsapi.PodDefaultList{}
-	err := crdclient.List(context.TODO(), &client.ListOptions{Namespace: pod.Namespace}, list)
+	err := crdclient.List(context.TODO(), list, &client.ListOptions{Namespace: pod.Namespace})
```
Comment (alexlatchford) [inline components/admission-webhook/main.go:486]: Dumb question why was this reordering needed?

File: components/admission-webhook/manifests/base/crd.yaml [hunk 2]
```diff
@@ -8,46 +8,62 @@ spec:
...
-      properties:
-        apiVersion:
-          type: string
-        kind:
-          type: string
-        metadata:
-          type: object
-        spec:
+  versions: 
+    - name: v1alpha1
```
Comment (alexlatchford) [inline components/admission-webhook/manifests/base/crd.yaml:14]: Maybe a dumb question but why is this one `v1alpha1`?

**Review by alexlatchford — APPROVED** [review]
Definitely not my area of expertise but looks great to me @aaron-arellano, thanks for tackling this!

File: components/admission-webhook/manifests/base/crd.yaml (hunk 2 above)
Comment (aaron-arellano) [reply components/admission-webhook/manifests/base/crd.yaml:14]: cause the version of the CRD (poddefault) did not change, we only changed the k8s CRD version.

File: components/admission-webhook/main.go (hunk 1 above)
Comment (aaron-arellano) [reply components/admission-webhook/main.go:486]: the function signature changed. when moving to v1 the original ordering did not work. below is the signature.

https://github.com/kubernetes-sigs/controller-runtime/blob/3cb67228604239d3cd764b41719565bb4a46add6/pkg/client/client.go#L258

**Review by Arhell — COMMENTED** [review]
/ok-to-test

**Comment by kimwnasptd** [comment]
Thanks for the contribution @aaron-arellano and thank you for your time reviewing this @alexlatchford!

I'll also take a look on Tuesday to wrap the 1.22 effort up

**Comment by aaron-arellano** [comment]
@kimwnasptd any updates on getting this review? Thanks!

**Comment by aaron-arellano** [comment]
@kimwnasptd will this be getting a review soon? We are looking to merge the changes on our end ASAP once the contribution gets approved here. Thanks

**Review by kimwnasptd — COMMENTED** [review]
These are the only questions I had. I'll also compile and apply these to a K8s 1.22 cluster to verify.

@aaron-arellano one last request, could you also update the `Makefile` to:
1. Use the `TAG ?= $(shell git describe --tags --always)`
2. Rename `build-gcr` to `docker-build`
3. Add one more rule `docker-push` that just pushes the image

File: components/admission-webhook/README.md [hunk 3]
```diff
@@ -75,10 +75,19 @@ This specifies
 1. When there is a pod being created (see `rules`),
 1. call the webhook service `gcp-cred-webhook.default` at path `/add-cred` (see `clientConfig`)
 
+### admissionregistration.k8s.io/v1 default failurePolicy
+In adopting `admissionregistration.k8s.io/v1` for the `MutatingWebhookConfiguration` we accept the default value
+for `failurePolicy` to be `Fail` per [documentation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy). Upon testing this default feature it was discovered if the `AdmissionWebhook`'s `mutating-webhook-configuration` failed to mutate a pod then the pod would fail to start, its associated `Deployment` or `StatefulSet` would continually attempt to create the pod until the process that created the pod was terminated. This continuous failure to mutate the target pod may block other target pods from mutation
+until the failing process ends.
```
[the remark this answers is not in the document]
Comment (aaron-arellano) [reply components/admission-webhook/README.md:81]: yep. will clarify. Pretty much any pod that is targeted by the admission webhook for mutation is effected and will fail to start if there are configuration collisions or some other issue. Already running pods are not effected at all, again only pods being created after the policy change.

File: components/admission-webhook/manifests/base/crd.yaml [hunk 4]
```diff
@@ -8,46 +8,62 @@ spec:
...
+                desc:
+                  type: string
+                serviceAccountName:
+                  type: string
+                automountServiceAccountToken:
+                  type: boolean
+                env:
+                  items:
+                    type: object
+                    x-kubernetes-preserve-unknown-fields: true
```
[the remark this answers is not in the document]
Comment (aaron-arellano) [reply components/admission-webhook/manifests/base/crd.yaml:37]: if I did not add `x-kubernetes-preserve-unknown-fields: true` I noticed the fields would get pruned in our pod-defaults, we would end up having empty pod-defaults though they were not prior to the upgrade. It seems the pruning mechanism recognizes `string`, `boolean` and `integer` no problem but did not recognize `array` or `object`. Once I added these tags the pod-defaults were generated as normal.

**Comment by aaron-arellano** [comment]
@kimwnasptd should I also update the image url in the makefile to `public.ecr.aws/j1r0q0g6/`? I saw that was updated in the notebook controller v1.22 code change.

**Comment by kimwnasptd** [comment]
Just tested this locally, as well as in a K8s 1.22 cluster.

Thanks for the work @aaron-arellano!

/lgtm
/approve

**Pull request merged**